src/probe/probe.c

```diff
@@ -15,42 +24,551 @@
 }
 
 int probe_dma_add(uint32_t count, struct probe_dma *probe_dma)
 {
+	uint32_t i;
+	uint32_t j;
+	uint32_t stream_tag;
+	uint32_t first_free;
+	int err;
+
+	tracev_probe("probe_dma_add() count = %u", count);
+
+	if (!_probe) {
+		trace_probe_error("probe_dma_add() error: Not initialized.");
+
+		return -EINVAL;
+	}
+
+	for (i = 0; i < count; i++) {
+		tracev_probe("\tprobe_dma[%u] stream_tag = %u, dma_buffer_size = %u",
+			     i, probe_dma[i].stream_tag,
+			     probe_dma[i].dma_buffer_size);
+
+		first_free = CONFIG_PROBE_DMA_MAX;
+
+		/* looking for first free dma slot */
+		for (j = 0; j < CONFIG_PROBE_DMA_MAX; j++) {
+			stream_tag = _probe->inject_dma[j].stream_tag;
+
+			if (stream_tag == PROBE_DMA_INVALID) {
+				if (first_free == CONFIG_PROBE_DMA_MAX)
+					first_free = j;
+
+				continue;
+			}
+
+			if (stream_tag == probe_dma[i].stream_tag) {
+				trace_probe_error("probe_dma_add() error: Probe DMA %u already attached.",
+						  stream_tag);
+				return -EINVAL;
+			}
+		}
+
+		if (first_free == CONFIG_PROBE_DMA_MAX) {
+			trace_probe_error("probe_dma_add() error: Exceeded maximum number of DMAs attached = " META_QUOTE(CONFIG_PROBE_DMA_MAX));
+			return -EINVAL;
+		}
+
+		_probe->inject_dma[first_free].stream_tag =
+			probe_dma[i].stream_tag;
+		_probe->inject_dma[first_free].dma_buffer_size =
+			probe_dma[i].dma_buffer_size;
+
+		err = probe_dma_init(&_probe->inject_dma[first_free]);
+		if (err < 0) {
+			trace_probe_error("probe_dma_add() error: probe_dma_init() failed");
+			_probe->inject_dma[first_free].stream_tag =
+				PROBE_DMA_INVALID;
+			return err;
+		}
+	}
+
 	return 0;
 }
 
 int probe_dma_info(struct sof_ipc_probe_info_params *data, uint32_t max_size)
 {
+	uint32_t i = 0;
+	uint32_t j = 0;
+
+	tracev_probe("probe_dma_info()");
+
+	if (!_probe) {
+		trace_probe_error("probe_dma_info() error: Not initialized.");
+
+		return -EINVAL;
+	}
+
+	data->rhdr.hdr.size = sizeof(*data);
+
+	/* iterate over inject DMAs */
+	while (i < CONFIG_PROBE_DMA_MAX &&
+	       data->rhdr.hdr.size + sizeof(struct probe_dma) < max_size) {
+		/* save it if valid */
+		if (_probe->inject_dma[i].stream_tag != PROBE_DMA_INVALID) {
+			data->probe_dma[j].stream_tag =
+				_probe->inject_dma[i].stream_tag;
+			data->probe_dma[j].dma_buffer_size =
+				_probe->inject_dma[i].dma_buffer_size;
+			j++;
+			/* and increase reply header size */
+			data->rhdr.hdr.size += sizeof(struct probe_dma);
+		}
+
+		i++;
+	}
+
+	return 1;
+}
+
+static int is_probe_stream_used(uint32_t stream_tag)
+{
+	uint32_t i;
+
+	for (i = 0; i < CONFIG_PROBE_POINTS_MAX; i++) {
+		if (_probe->probe_points[i].stream_tag == stream_tag)
+			return 1;
+	}
+
 	return 0;
 }
 
 int probe_dma_remove(uint32_t count, uint32_t *stream_tag)
 {
+	uint32_t i;
+	uint32_t j;
+	int err;
+
+	tracev_probe("probe_dma_remove() count = %u", count);
+
+	if (!_probe) {
+		trace_probe_error("probe_dma_remove() error: Not initialized.");
+
+		return -EINVAL;
+	}
+
+	for (i = 0; i < count; i++) {
```

Comment:
what are we doing here.

Reply:
I think its obvious: "probe_dma_remove(uint32_t count, uint32_t *stream_tag)" but added small comment